### PR TITLE
fix: iterative flattening pass

### LINKED ISSUE #4449: Stack overflow when having loops containing if statements
### Problem

The PR https://github.com/AztecProtocol/aztec-packages/pull/4571 is failing CI because it reaches a stack overflow during the ssa flatenning.
I managed to get a minimal repro with the following code:
```
fn main(x: u8, result: [u8; 32]) {
    let mut digest = [0;32];
    for i in 0..70 {
        let y = x+i;
    let a = [y ,x,32,0,y+1,y-1,2*y,5];
    digest = std::sha256::digest(a);
    }

    assert(digest == result);
}
```

### Happy Case

The flattening pass is using recursion and should be changed into an iterative version

### Project Impact

None

### Impact Context

_No response_

### Workaround

None

### Workaround Description

_No response_

### Additional Context

_No response_

### Would you like to submit a PR for this Issue?

Yes

### Support Needs

_No response_

### PR DESCRIPTION
**Pull request by guipublic**
# Description

## Problem\*

Resolves #4449 

## Summary\*

The issue is fixed by implementing an iterative version of the flattening pass.

## Documentation\*

Check one:
- [X] No documentation needed.
- [ ] Documentation included in this PR.
- [ ] **[Exceptional Case]** Documentation to be submitted in a separate PR.

# PR Checklist\*

- [X] I have tested the changes locally.
- [X] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: compiler/noirc_evaluator/src/ssa/opt/flatten_cfg.rs [hunk 1]
```diff
@@ -233,90 +258,121 @@ fn flatten_function_cfg(function: &mut Function) {
...
+        // until all blocks have been flattened.
+        // We follow the terminator of each block to determine which blocks to
+        // process next
+        let mut queue = vec![self.inserter.function.entry_block()];
+        while let Some(block) = queue.pop() {
+            self.inline_block(block);
+            let to_process = self.handle_terminator(block, &queue);
+            for i in to_process {
+                if !queue.contains(&i) {
+                    queue.push(i);
```
Comment (jfecher) [inline compiler/noirc_evaluator/src/ssa/opt/flatten_cfg.rs:280]:
```suggestion
            for incoming_block in to_process {
                if !queue.contains(&incoming_block) {
                    queue.push(incoming_block);
```

File: compiler/noirc_evaluator/src/ssa/opt/flatten_cfg.rs [hunk 2]
```diff
@@ -233,90 +258,121 @@ fn flatten_function_cfg(function: &mut Function) {
...
+    /// Returns the current condition
+    fn get_last_condition(&self) -> Option<ValueId> {
+        if let Some(context) = self.condition_stack.last() {
+            if let Some(else_branch) = &context.else_branch {
+                return Some(else_branch.condition);
+            } else {
+                return Some(context.then_branch.condition);
+            }
+        }
+        None
```
Comment (jfecher) [inline compiler/noirc_evaluator/src/ssa/opt/flatten_cfg.rs:308]:
```suggestion
        self.condition_stack.last().map(|context| {
            match &context.else_branch {
                Some(else_branch) => else_branch.condition,
                None => context.then_branch.condition,
            }
        })
```

**Review by jfecher — APPROVED** [review]
LGTM

**Comment by guipublic** [comment]
In the resolved issue, there is a regression test that I can easily add. However I am hesitant to add a regression test for this because it might affect performance of the CI.

**Comment by vezenovm** [comment]
> In the resolved issue, there is a regression test that I can easily add. However I am hesitant to add a regression test for this because it might affect performance of the CI.

If the circuit is too big we can hold off. How large is it?

**Pull request merged**